File: apps/desktop/src/renderer/lib/byokProviderPresets.ts
/**
 * BYOK vendor presets — frontend-only catalog for model settings.
 *
 * Each preset supplies a canonical base_url and common model IDs; users may
 * still override model names (their key determines what actually works).
 */

export type ByokProviderId =
  | "openai"
  | "deepseek"
  | "moonshot"
  | "zhipu"
  | "doubao"
  | "openrouter"
  | "custom";

export interface ByokProviderPreset {
  id: Exclude<ByokProviderId, "custom">;
  label: string;
  baseUrl: string;
  /** Alternate base URLs that should map to this preset (e.g. regional endpoints). */
  baseUrlAliases?: readonly string[];
  defaultModel: string;
  models: readonly string[];
  keyHelpUrl?: string;
}

export const BYOK_CUSTOM_PROVIDER_ID = "custom" as const;

export const BYOK_PROVIDER_PRESETS: readonly ByokProviderPreset[] = [
  {
    id: "deepseek",
    label: "DeepSeek",
    baseUrl: "https://api.deepseek.com",
    baseUrlAliases: ["https://api.deepseek.com/v1"],
    defaultModel: "deepseek-v4-flash",
    models: ["deepseek-v4-flash", "deepseek-v4-pro"],
    keyHelpUrl: "https://platform.deepseek.com/api_keys",
  },
  {
    id: "openai",
    label: "OpenAI",
    baseUrl: "https://api.openai.com/v1",
    defaultModel: "gpt-4o",
    models: ["gpt-4o", "gpt-4o-mini", "o3-mini"],
    keyHelpUrl: "https://platform.openai.com/api-keys",
  },
  {
    id: "moonshot",
    label: "Kimi (Moonshot)",
    baseUrl: "https://api.moonshot.cn/v1",
    baseUrlAliases: ["https://api.moonshot.ai/v1"],
    defaultModel: "kimi-k2.5",
    models: ["kimi-k2.5", "kimi-k2", "moonshot-v1-8k", "moonshot-v1-32k"],
    keyHelpUrl: "https://platform.moonshot.cn/console/api-keys",
  },
  {
    id: "zhipu",
    label: "智谱 GLM",
    baseUrl: "https://open.bigmodel.cn/api/paas/v4",
    defaultModel: "glm-4-plus",
    models: ["glm-4-plus", "glm-4-flash", "glm-4-air"],
    keyHelpUrl: "https://open.bigmodel.cn/usercenter/apikeys",
  },
  {
    id: "doubao",
    label: "豆包 (火山方舟)",
    baseUrl: "https://ark.cn-beijing.volces.com/api/v3",
    defaultModel: "doubao-pro-32k",
    models: ["doubao-pro-32k", "doubao-lite-32k"],
    keyHelpUrl:
      "https://console.volcengine.com/ark/region:ark+cn-beijing/apiKey",
  },
  {
    id: "openrouter",
    label: "OpenRouter",
    baseUrl: "https://openrouter.ai/api/v1",
    defaultModel: "openrouter/auto",
    models: [
      "openrouter/auto",
      "anthropic/claude-sonnet-4",
      "google/gemini-2.5-pro",
    ],
    keyHelpUrl: "https://openrouter.ai/keys",
  },
] as const;

export const DEFAULT_BYOK_PROVIDER_ID: Exclude<ByokProviderId, "custom"> =
  "deepseek";

const PRESET_BY_ID = new Map(
  BYOK_PROVIDER_PRESETS.map((preset) => [preset.id, preset]),
);

/** Normalize base_url for preset matching (case, trailing slashes). */
export function normalizeByokBaseUrl(url: string): string {
  let normalized = url.trim().toLowerCase();
  while (normalized.endsWith("/")) {
    normalized = normalized.slice(0, -1);
  }
  return normalized;
}

function presetBaseUrls(preset: ByokProviderPreset): string[] {
  return [preset.baseUrl, ...(preset.baseUrlAliases ?? [])];
}

export function getByokProviderPreset(
  id: Exclude<ByokProviderId, "custom">,
): ByokProviderPreset {
  const preset = PRESET_BY_ID.get(id);
  if (!preset) {
    throw new Error(`Unknown BYOK provider preset: ${id}`);
  }
  return preset;
}

export function isCustomByokProvider(
  id: ByokProviderId,
): id is typeof BYOK_CUSTOM_PROVIDER_ID {
  return id === BYOK_CUSTOM_PROVIDER_ID;
}

/** Match stored base_url to a preset, or fall back to custom. */
export function resolveByokProviderFromConfig(baseUrl: string): ByokProviderId {
  const trimmed = baseUrl.trim();
  if (!trimmed) return DEFAULT_BYOK_PROVIDER_ID;

  const normalized = normalizeByokBaseUrl(trimmed);
  for (const preset of BYOK_PROVIDER_PRESETS) {
    if (
      presetBaseUrls(preset).some(
        (candidate) => normalizeByokBaseUrl(candidate) === normalized,
      )
    ) {
      return preset.id;
    }
  }
  return BYOK_CUSTOM_PROVIDER_ID;
}

export function listByokProviderOptions(): Array<{
  id: ByokProviderId;
  label: string;
}> {
  return [
    ...BYOK_PROVIDER_PRESETS.map((preset) => ({
      id: preset.id,
      label: preset.label,
    })),
    { id: BYOK_CUSTOM_PROVIDER_ID, label: "自定义" },
  ];
}

File: apps/server/agentcore/llm/byok_provider_presets.py
"""BYOK vendor presets — server-side catalog seed aligned with desktop.

Mirrors ``apps/desktop/src/renderer/lib/byokProviderPresets.ts`` (baseUrl / aliases /
models / defaultModel). Catalog merge matches providers by normalized ``base_url``;
unknown endpoints get no preset rows.
"""

from __future__ import annotations

from dataclasses import dataclass


@dataclass(frozen=True)
class ByokProviderPreset:
    id: str
    label: str
    base_url: str
    default_model: str
    models: tuple[str, ...]
    base_url_aliases: tuple[str, ...] = ()


BYOK_PROVIDER_PRESETS: tuple[ByokProviderPreset, ...] = (
    ByokProviderPreset(
        id="deepseek",
        label="DeepSeek",
        base_url="https://api.deepseek.com",
        base_url_aliases=("https://api.deepseek.com/v1",),
        default_model="deepseek-v4-flash",
        models=("deepseek-v4-flash", "deepseek-v4-pro"),
    ),
    ByokProviderPreset(
        id="jiurelay",
        label="JiuRelay",
        base_url="https://jiurelay.com/openai/v1",
        default_model="glm-5.2",
        models=("glm-5.2", "deepseek-v4-flash-0731", "grok-4.5"),
    ),
    ByokProviderPreset(
        id="openai",
        label="OpenAI",
        base_url="https://api.openai.com/v1",
        default_model="gpt-4o",
        models=("gpt-4o", "gpt-4o-mini", "o3-mini"),
    ),
    ByokProviderPreset(
        id="moonshot",
        label="Kimi (Moonshot)",
        base_url="https://api.moonshot.cn/v1",
        base_url_aliases=("https://api.moonshot.ai/v1",),
        default_model="kimi-k2.6",
        # kimi-k2 / moonshot-v1-* retired; k2.5 kept for older keys.
        models=("kimi-k2.6", "kimi-k3", "kimi-k2.5"),
    ),
    ByokProviderPreset(
        id="zhipu",
        label="智谱 GLM",
        base_url="https://open.bigmodel.cn/api/paas/v4",
        default_model="glm-4-plus",
        models=("glm-4-plus", "glm-4-flash", "glm-4-air"),
    ),
    ByokProviderPreset(
        id="doubao",
        label="豆包 (火山方舟)",
        base_url="https://ark.cn-beijing.volces.com/api/v3",
        default_model="doubao-seed-2-1-turbo-260628",
        # Short seed; doubao-pro/lite-32k retired — use dated seed IDs or ep-… endpoints.
        models=("doubao-seed-2-1-turbo-260628",),
    ),
    ByokProviderPreset(
        id="hy",
        label="腾讯 Hy (TokenHub)",
        base_url="https://tokenhub.tencentmaas.com/v1",
        base_url_aliases=(
            "https://tokenhub.tencentmaas.cn/v1",
            "https://tokenhub-intl.tencentmaas.com/v1",
            "https://tokenhub-intl.tencentmaas.cn/v1",
        ),
        default_model="hy3",
        models=("hy3", "hy3-preview"),
    ),
    ByokProviderPreset(
        id="openrouter",
        label="OpenRouter",
        base_url="https://openrouter.ai/api/v1",
        default_model="openrouter/auto",
        models=(
            "openrouter/auto",
            "anthropic/claude-sonnet-4",
            "google/gemini-2.5-pro",
        ),
    ),
    ByokProviderPreset(
        id="opencode_zen",
        label="OpenCode Zen",
        base_url="https://opencode.ai/zen/v1",
        default_model="deepseek-v4-flash",
        # Short seed for discovery-miss; full catalog = GET /models union.
        models=("deepseek-v4-flash", "kimi-k2.6", "glm-5.2"),
    ),
)


def normalize_byok_base_url(url: str) -> str:
    """Normalize base_url for preset matching (case, trailing slashes)."""
    normalized = url.strip().lower()
    while normalized.endswith("/"):
        normalized = normalized[:-1]
    return normalized


def _preset_base_urls(preset: ByokProviderPreset) -> tuple[str, ...]:
    return (preset.base_url, *preset.base_url_aliases)


def match_byok_provider_preset(base_url: str) -> ByokProviderPreset | None:
    """Return the preset whose canonical / alias base_url matches, else None."""
    normalized = normalize_byok_base_url(base_url)
    if not normalized:
        return None
    for preset in BYOK_PROVIDER_PRESETS:
        if any(
            normalize_byok_base_url(candidate) == normalized
            for candidate in _preset_base_urls(preset)
        ):
            return preset
    return None


def preset_models_for_base_url(base_url: str) -> tuple[str, ...]:
    """Model ids from the matching vendor preset, or empty when unknown/custom."""
    preset = match_byok_provider_preset(base_url)
    return preset.models if preset is not None else ()
